analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs

```diff
@@ -255,6 +255,9 @@ static SyntaxNode PushPathPositionForTuple(Stack<PathPosition> pathPositions, Tu
                     : null;
         }
 
+        public static bool IsInTopLevelMain(this SyntaxNode node) =>
+            node.SyntaxTree.GetRoot()?.ChildNodes().OfType<GlobalStatementSyntax>().Any() ?? false;
```

Comment:
Looks valid to me. 
Have you considered this:
It is unclear to me why it returns `true` if wrapped in a method in a top-level main. This is not compatible with `context.ContainingSyntax` used before. You may consider providing two extensions methods instead:
* One for SyntaxTree
* One for node, this one returns `true` only if the Ancestors don't have any method declaration or type declaration but a `SyntaxKind.GlobalStatement`.

If you want to keep it as is, you should document the behavior with a doc comment, and also, `ChildNodes()` can be avoided. Some optimizations you may consider:
`node.SyntaxTree.GetRoot() is { CompilationUnit { Members: { Length: >0 } members } } && members.Any(x => x.IsKind(SyntaxKind.GlobalStatement)`
`node.SyntaxTree.GetRoot().ChildNodes().Any(x => x.IsKind(SyntaxKind.GlobalStatement)`
or a combination of it.

Reply:
`node is CompilationUnitSyntax compilationUnit && compilationUnit.Members.Any(x => x.IsKind(SyntaxKind.GlobalStatement));` does the trick for all rules but one: `MethodsShouldNotHaveTooManyLines`. In this case, we subscribe for local functions and we want to handle also the ones from top level main.

Reply:
> and we want to handle also the ones from top level main.

So you just don't call `IsInTopLevelMain` then?

Reply:
Yes, I'm thinking to change the check a bit but only for `MethodsShouldNotHaveTooManyLines`